python/cugraph/cores/k_core_wrapper.pyx

```diff
@@ -28,83 +28,101 @@ from libc.float cimport FLT_MAX_EXP
+    df['dst'] = cudf.Series(data=dst, dtype="int32")
+    if weight_type(input_graph) == np.float32:
+        wgt = Buffer(wgt)
+        df['weight'] = cudf.Series(data=wgt, dtype="float32")
     
-    if weights_type == np.float32:
-        in_graph_float = GraphCOO[int,int,float](<int*>c_src, <int*>c_dst, <float*>c_weights, num_verts, num_edges)
-        c_k_core[int,int,float](in_graph_float, k, <int*> c_vertex, <int*> c_values, len(core_number), out_graph_float)
-
-        tmp = rmm.device_array_from_ptr(<uintptr_t>out_graph_float.src_indices,
-                                        nelem=out_graph_float.number_of_edges,
-                                        dtype=np.int32)
-        df['src'] = cudf.Series(tmp)
-
-        tmp = rmm.device_array_from_ptr(<uintptr_t>out_graph_float.dst_indices,
-                                        nelem=out_graph_float.number_of_edges,
-                                        dtype=np.int32)
-        df['dst'] = cudf.Series(tmp)
-
-        if weights is not None:
-            tmp = rmm.device_array_from_ptr(<uintptr_t>out_graph_float.edge_data,
-                                            nelem=out_graph_float.number_of_edges,
-                                            dtype=np.int32)
-            df['weight'] = tmp
-    else:
-        in_graph_double = GraphCOO[int,int,double](<int*>c_src, <int*>c_dst, <double*>c_weights, num_verts, num_edges)
-        c_k_core[int,int,double](in_graph_double, k, <int*> &c_vertex, <int*> &c_values, len(core_number), out_graph_double)
-
-        tmp = rmm.device_array_from_ptr(<uintptr_t>out_graph_double.src_indices,
-                                        nelem=out_graph_double.number_of_edges,
-                                        dtype=np.int32)
-        df['src'] = cudf.Series(tmp)
-
-        tmp = rmm.device_array_from_ptr(<uintptr_t>out_graph_double.dst_indices,
-                                        nelem=out_graph_double.number_of_edges,
-                                        dtype=np.int32)
-        df['dst'] = cudf.Series(tmp)
-
-        if weights is not None:
-            tmp = rmm.device_array_from_ptr(<uintptr_t>out_graph_double.edge_data,
-                                            nelem=out_graph_double.number_of_edges,
-                                            dtype=np.int32)
-            df['weight'] = tmp
-        
+    return df
 
+
+def k_core_double(input_graph, k, core_number):
+    c_src, c_dst, c_weights, num_verts, num_edges = graph_params(input_graph)
+    c_vertex, c_values = core_number_params(core_number)
+
+    cdef GraphCOOView[int,int,double] in_graph
+    in_graph = GraphCOOView[int,int,double](<int*>c_src, <int*>c_dst, <double*>c_weights, num_verts, num_edges)
+    cdef unique_ptr[GraphCOO[int,int,double]] out_graph = move(c_k_core[int,int,double](in_graph, k, <int*>c_vertex, <int*>c_values, len(core_number)))
+    cdef GraphCOOContents[int,int,double] contents = move(out_graph.get()[0].release())
```

Comment:
This seems like it will be a common idiom for things that return a graph.

Would it make sense to add a function to graph_new_wrapper.pyx that takes: `cdef GraphCOO[*]` and returns a cudf data frame (or a triple of cudf.Series objects: src, dst, weight)?  That way k-core, subgraph extraction, subgraph matching (once gunrock gets it working), and any other functions that generate a graph can call it?

Might also wants one that takes GraphCSR[*] and returns cudf.Series for offsets, indices, weights.

I don't know a way to "template" them, so we might need to create all variations.

Reply:
It does seem like the results of function that return a graph (GraphCSR or GraphCOO) that a user would then not be able to pass that graph into another algorithm

Reply:
You can't pass the GraphCSR or GraphCOO into another algorithm.  You would first convert the resulting GraphCSR/GraphCOO into a python graph object (python object takes ownership of the pointers).  That's what this code is doing.  Technically the wrapper is creating a Dataframe with the GraphCOO in it.  The k_core.py takes this data frame and creates a python graph object from it.  The python graph object could be used to call another algorithm.